File: usacobronze/lineup.py
with open("lineup.in","r") as fin:
    ins=fin.readlines()
n=int(ins.pop(0))
# first assume that the input is in order
# res=["*", "Bessie","*", "Buttercup", "*", "Belinda", "*", "Beatrice", "*", "Bella", "*", "Blue", "*", "Betsy", "*","Sue","*"]
# do some sorting to the ins to have the names appear in order
a=[]
for x in ins:
    y=x.split()
    a.extend((y[0],y[-1]))
nnum={"Bessie":0,"Buttercup":1,"Belinda":2,"Beatrice":3,"Bella":4,"Blue":5,"Betsy":6,"Sue":7}
ins=[[nnum[x.split()[0]],nnum[x.split()[-1]]] for x in ins]
# ord=[ins.pop(0)]
# for x in ins:
#
#     if x[0]==ord[-1][-1]:
#         ord[-1].append(x[1])
#     elif x[1]==ord[-1][-1]:
#         ord[-1].append(x[0])
#     else:
#         ord.append(x)
# print(ord)
# dict recursion again lmao
print(ins)
nextto={x:[] for x in range(8)}
for x in ins:
    nextto[x[0]].append(x[1])
    nextto[x[1]].append(x[0])
print(nextto)
# every value either has 0,1, or 2 values
# each key can also only appear at most twice in any other value
freemove=[]
groups=[[]]
def getseq(cow):
    # init cow must be the start of the seq
    if not nextto[cow]:
        return [cow]
    if len(nextto[cow])==1:
        # it's on the edge
        return [cow]+getseq(nextto[cow][0])
    for x in nextto[cow]:
        if x==cow:
            continue
        return [cow]+getseq(x)
    # must be surrounded then

print(groups)
try:
    for x in nextto:
        if not nextto[x]:
            freemove.append(x)
            continue
        if len(nextto[x])==1:
            # start/end of a sequences
            groups.append(getseq(x))
except: pass
# ins.pop()
# res=[a.pop(),a.pop()]
# print(res)
# look at index 1 and 2, then 3 and 4, etc. If the two indexes are the same, then add that to the group that already
# exists, otherwise make a new group.
groups=[[a[0]]]
for x in range(1,len(a)-1,2):
    if a[x]==a[x+1]:
        groups[-1].extend((a[x],a[x+2]))
    else:
        groups.append([a[x]])
# res=[[]]
# for x in range(2,len(a)-1,2):
#     if a[x]==a[x-1]:
#         res[-1].extend((a[x-2],a[x+1]))
#     else:
#         res.append(a[x:x+2])
# print(res)
# print(a)
# group=[list(ins.pop())]
# print(group)
# for x in ins:
#     found=False
#     for groups in group:
#         if x[0] in groups:
#             found=True
#             if groups.index(x[0])==0:
#                 groups.insert(0,x[1])
#             else:
#                 groups.append(x[1])
#             ins.remove(x)
#
#         if x[1] in groups:
#             found=True
#             if groups.index(x[1])==0:
#                 groups.insert(0,x[0])
#             else:
#                 groups.append(x[0])
#             ins.remove(x)
#     if not found:
#         group.append(list(x))
#         ins.remove(x)
#     print(x)
# print(group)
# consider=a.pop()
# for x in range(len(a)):
#     print(x)

# for x in range(len(a)):
#     if a.count(a[x])>1:
#         if x%2: # x is in an odd index


with open("lineup.out","w") as fout:
    pass

File: usacobronze/lineup.cpp
#include <bits/stdc++.h>
using namespace std;
#define pss pair<str,str>
#define vt vector
#define str string
#define f first
#define s second
#define b begin()
#define e end()
#define gb groups.back()

vt<str> half1,half2; // reverse half2 when combining
void strip_ends(vt<str> group, str adj_left, str adj_right) {
    // what is originally to the left and right of the list, these numbers would be in group
    // the base case would look like {a,b,b,c,c,d} or {a,b,b,c}
    // base case 1 can be reduced to {b,c,c,d} after adding b to half1
    // this works for cow lines of length 3 and above.
    // length 2 is an edge case
    for (str i: group) cout << i << "\" \"";
    cout << endl;
    if (group.size()==4) {
        for (int i=0;i<4;i++) {
            if (group[i]==adj_left) {
                // if the index is even, then add the next one, else add the previous one
                half1.push_back(group[i+(i%2? 1:-1)]);
            }
            if (group[i]==adj_right) {
                half2.push_back(group[i+(i%2? 1:-1)]);
            }
        }
        return;
    }
    if (group.size()==6) {
        for (int i=0;i<6;i++) {
            if (group[i]==adj_left) {
                half1.push_back(group[i+(i%2? 1:-1)]);
                group.erase(group.begin()+i);
                group.erase(group.begin()+i+(i%2? 0:-1));
                strip_ends(group,half1.back(),adj_right);
                return;
            }
        }
    }
    // find the two ends, cut them off and carry on
    str new_left="-";
    str new_right="-";
    for (int i=0;i<group.size();i++) {
        if (group[i]==adj_left) {
            half1.push_back(group[i+(i%2? 1:-1)]);
            new_left=half1.back();
            group.erase(group.begin()+i);
            group.erase(group.begin()+i+(i%2? 0:-1));
        }
        if (group[i]==adj_right) {
            half2.push_back(group[i+(i%2? 1:-1)]);
            new_right=half2.back();
            group.erase(group.begin()+i);
            group.erase(group.begin()+i+(i%2? 0:-1));
        }
    }
    assert (new_left!="-" && new_right!="-");
    strip_ends(group,new_left,new_right);
    

}
int main() {
    ifstream cin("lineup.in");
    // ofstream cout("lineup.out");
    int n;
    cin >> n;
    vt<pss> beside(n);
    str raw_input;
    for (int i=0;i<n;i++) {
        cin >> raw_input;
        beside[i].f=raw_input;
        cin >> raw_input >> raw_input >> raw_input >> raw_input >> raw_input; // get the fifth word after the first name, alternatively use cin.ignore()
        beside[i].s=raw_input;
    }

    // among each group of adjacent cows there would only be one cow name present on each edge. Max 3 groups
    vt<vt<str> > groups={{beside.back().f,beside.back().s}};
    beside.pop_back();
    // we can just look through, and if there are none to add then pop back the next one - the list is short enough that this is alright

    while (!beside.empty()) {
        // look at each pair and see if it's in the back group
        vt<int> to_remove;
        for (int i=0;i<beside.size();i++) {
            pss p=beside[i];
            if (count(gb.b,gb.e, p.f)==1) {
                // the first elem appears in the back group. Add the second one and the first to know where the ends are
                gb.push_back(p.f);
                gb.push_back(p.s);
                to_remove.push_back(i);
            }
            else if (count(gb.b,gb.e,p.s)==1) {
                // same story
                gb.push_back(p.f);
                gb.push_back(p.s);
                to_remove.push_back(i);
            }
        }
        if (to_remove.empty()) {
            // the group has finished
            groups.push_back({beside.back().f,beside.back().s});
            beside.pop_back();
        }
        for (int i=0;i<to_remove.size();i++) {
            beside.erase(beside.b+to_remove[i]-i); // big brain - account for the fact that size decreases by 1 every time
        }
    }
    vt<vt<str> > res;
    for (vt<str> v: groups) {

        // v.insert(v.begin(),v.front());
        // v.insert(v.begin(),"a");
        // v.push_back(v.back());
        // v.push_back("a");
        if (v.size()==2) {
            res.push_back(v);
            continue;
        }
        map<str,int> appear;
        for (auto i: v) appear[i]++;
        vt<str> edges;
        for (auto i: v) if (appear[i]==1) {
            edges.push_back(i);
        }
        assert (edges.size()==2);
        strip_ends(v,edges[0],edges[1]);
        res.resize(res.size()+1);
        for (int i=0;i<half2.size();i++) {
            res.back().push_back(half1[i]);
            res.back().push_back(half2[half2.size()-i]);
        }
        if (half1.size()!=half2.size()) {
            res.back().push_back(half1.back());
        }

        half1.clear();
        half2.clear();

    }
    for (auto i: res) {
        for (auto j: i) cout << j << endl;
        cout << endl;
    }
    // const str alpha_names[] = {"Bessie", "Buttercup", "Belinda", "Beatrice", "Bella", "Blue", "Betsy", "Sue"};
    // for (auto v: groups) {
    //     assert (v.size()%2==0);
    //     vt<str> new_group;
    //     // find a string that appears once - there should be two
    //     // get the original back name, use to avoid confusion later and add to new_group
    //     map<str,int> appear;
    //     for (str s: v) appear[s]++;
    //     str orig_end="\n";
    //     for (int i=v.size()-1;i>-1;i--) {
    //         if (appear[v[i]]==1) {
    //             orig_end=v[i];
    //             break;
    //         }
    //     }
    //     assert (orig_end!="\n");
    //     // watch out for the single group edge case, see if this works first though
    //     while (v.size()>1) {
    //         map<str,int> appear;
    //         for (str s: v) {
    //             appear[s]++;
    //         }
    //         for (int i=0;i<v.size();i++) {
    //             if (appear[v[i]]==1 && v[i]!=orig_end) {
    //                 new_group.push_back(v[i]);
    //                 new_group.push_back(v[i+1]);
    //                 // delete those two from the array and carry on
    //                 v.erase(v.begin()+i);
    //                 v.erase(v.begin()+i); // previous index doesn't exist enymore
    //                 break;
    //             }
    //         }
    //     }
    //     new_group.push_back(orig_end);
    //     std::cout << "one iter done";
    // }
    
}